packages/passport-interface/src/SubscriptionManager.ts

```diff
@@ -61,15 +61,102 @@ export class FeedSubscriptionManager {
   public constructor(
     api: IFeedApi,
     providers?: SubscriptionProvider[],
-    activeSubscriptions?: Subscription[]
+    activeSubscriptions?: Subscription[],
+    errors?: Map<string, SubscriptionError>
   ) {
     this.updatedEmitter = new Emitter();
     this.api = api;
     this.providers = providers ?? [];
     this.activeSubscriptions = (activeSubscriptions ?? []).map(ensureHasId);
-    this.errors = new Map();
+    this.errors = errors !== undefined ? new Map(errors) : new Map();
   }
 
+  /**
+   * Creates a new FeedSubscriptionManager with the data from this one.
+   *
+   * This includes all data (subscriptions, providers, and errors, but doesn't
+   * include dynamic state (listeners on the emitter).
+   *
+   * This is a shallow clone.  The resulting object has new containers but
+   * the same underlying objects, which are expected to be immutable.  If you
+   * need a deep clone, use serialization.
+   */
+  public clone(): FeedSubscriptionManager {
+    return new FeedSubscriptionManager(
+      this.api,
+      [...this.providers],
+      [...this.activeSubscriptions],
+      this.errors
+    );
+  }
+
+  /**
+   * Merges subscriptions from `other` into `this`, returning an object with
+   * a count of the number of new providers and subscriptions added.
+   *
+   * Only subscriptions and providers are included in the merge.  Other
+   * non-persistent state (listeners, errors) is ignored.
+   *
+   * The merge only ever adds new providers and/or subscriptions which do not
+   * already exist.  Existing entries are never mutated.  Subscriptions will be
```

Comment:
This function isn't called yet, but you're correct about where it's intended to be called.  The choice of rules here reflects the same set of assumptions and priorities for PCDs.  Specifically, when there's doubt about whether or not an object should be in our store, we always prefer to keep it.  Also we treat the individual objects as atomic blobs, and in the case of conflict we have to pick one of them not merge individual fields.  What I'm taking away from your thoughts here is that "prefer to keep" is less obviously preferable here due to the security implications.  Note that the case without merge isn't "prefer to delete".  Without merge, if a conflict happens the outcome is entirely unpredictable.

I take your point that there's a security aspect to this which needs to be considered.  I do think it's a bit tricky to consider all future uses of permissions within the context of the current merge effort which is very limited and targeted to solve today's problems.  From a user's point of view in current use cases, subscriptions are something they either have or don't have, so the "prefer to keep" logic seems applicable.  E.g. if I claimed a new frog biome via a QR code, I'll be upset if it disappears.  However there is some level of destructiveness which makes it less obvious.  I don't think there are any actual examples of feeds today where unsubscribing has useful results, primarily because all the ticket feeds get re-added by default.  But I can imagine cases within the current set of supported features.  E.g. maybe I unsubscribe from a ticket feed because I don't want updates to my ticket after I've made proofs about it, or I don't want a revoked version.

I'd like to better understand the scope of what can go wrong here today.  E.g. do we actually support any "read" permissions which would upload PCDs to a feed?  I haven't seen such a case, other than the semaphore signature credential.  I think the only "attack" here today is to allow updating/deleting of PCDs in a specific folder.  I can imagine an attacker tricking a user into giving their feed the ability to delete in the root folder, in which case the user having difficulty removing the feed could be problematic.

We should probably talk about this more interactively.  We could decide not to merge subscriptions at all.  If we don't merge them, then new feeds might disappear, and removed feeds might reappear, at similar probability,   If we include this kind of merge, new feeds should never disappear, but removed feeds are more likely to reappear.  I think that's better for non-malicious use cases, but don't have a great handle on how real the malicious use cases are likely to be.

Reply:
If merging subscriptions interactively during an import, we could ask the user to confirm any permissions that imported subscriptions would have. Sync is the special case where we implicitly trust that the data being merged in must have been approved by the user.

Reply:
Right, what we're missing is the time aspect.  Approved by the user at some point, but they might have intended to revoke approval.  We can't solve that properly without getting into a more complicated form of merge (3-way merge with a base rev), and/or keeping extra data (tombstones, revision history), all of which are reasonable possibilities, but not this week.